### ISSUE TITLE
[Bug]: plugin-transform-block-scoping constant checks treats deconstruction incorrectly

### ISSUE THREAD

**Issue by overlookmotel** [babel/babel#13252]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

@babel/cli

### Input code

```js
const c = 0;
let x = 1, y = 2;
try {
  [x, c, y] = [11, 0, 22];
} catch (e) {
  console.log( { x, y } );
}
```

[REPL](https://babeljs.io/repl#?browsers=&build=&builtIns=false&corejs=3.6&spec=false&loose=true&code_lz=MYewdgzgLgBMMF4YAYDcAoANgU1gD0RgEYAaGAT0ICYMoAnSgb3RhgG08zgzyBdQtkVIoyVKrwwBfOAEMowABYwAFNgCUMZq1CQQOAHSYQAc2WNOFSWqnogA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=&prettier=false&targets=Node-10&version=7.14.0&externalPlugins=%40babel%2Fplugin-transform-block-scoping%407.13.16) 

### Current and expected behavior

`[x, c, y] = [11, 0, 22]` is transformed to `[11, 0, 22], _readOnlyError("c")`.

Original logs `{x: 11, y: 2}` whereas transpiled code logs `{x: 1, y: 2}`. It should assign to `x` before throwing, but not assign to `c` or `y`.

### Environment

Babel REPL

### Possible solution

Best solution I can see is:

```js
// Assignments to `c` and `y` have been removed
[x] = [11, 0, 22], _readOnlyError("c");
```

If there's a nested object/array on the receiving end of the assignment, all elements before the const-violating identifier should be retained, and all after it removed - at all levels of the nested object/array.

i.e.: if `c` is a const:

Input: `[ p, q, { r: [ ...s ], t: [ u, c, v ], w, ...x }, y, ...z ] = arr;`
Output: `[ p, q, { r: [ ...s ], t: [ u ] } ] = arr, _readOnlyError("c");`

(note `c`, `v`, `w`, `x`, `y` and `z` have all been removed)

NB `c` may also have a default e.g. `{ c = 1 } = {}`

### Additional context

#13248 fixed some other cases where const violations were transformed incorrectly, but didn't fix this one.

**Comment by overlookmotel** [comment]
Actually, solution is a bit more complex due to getters. e.g.:

```js
let getterIsCalled = false;
const obj = {
  get x() {
    getterIsCalled = true;
  }
};

try {
  const c = 1;
  ( { x: c } = obj );
} catch (e) {
  console.log( getterIsCalled );
}
```

This logs "true".

So, to make sure getter is called before throwing, transformation would need to be:

```js
var c = 1;
var _c;
( { x: _c } = obj ), _readOnlyError("c");
```

**Comment by overlookmotel** [comment]
Same bug with imports which are assigned to in object/array deconstruction in [plugin-transform-modules-commonjs](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs). e.g.: this is not transformed correctly:

```js
import c from 'foo';
let x = 0;
[ x, c ] = [ 1, 2 ];
```

**Comment by overlookmotel** [comment]
There is one rather ugly solution which would be easier to implement.

Input:

```js
const c = 1;
let x = 0, y = 0;
( { x, c, y } = { x: 1, c: 2, y: 3 } );
```

Output:

```js
function _readOnlyErrorObject(name) {
  return {
    set a(v) {
       _readOnlyError(name);
    }
  };
}

var c = 1;
var x = 0, y = 0;
[ x, _readOnlyErrorObject("c").a, y ] = [ 1, 2, 3 ];
```

This does behave correctly - the error only gets thrown when the setter is triggered, so `x` is assigned to, `y` is not, and any getter on the `c` property of right-hand object is called before throwing.

If you'd be willing to accept such an ugly solution, I'd be willing to make a PR.

**Comment by nicolo-ribaudo** [comment]
Would reusing the existing helper work?
```js
var c = 1;
var x = 0, y = 0;
[ x, _readOnlyError("c")._, y ] = [ 1, 2, 3 ];
```

**Comment by overlookmotel** [comment]
@nicolo-ribaudo No. I tried that. Problem is that `_readOnlyError("c")` gets evaluated before the property getter on object on right-hand side gets called. So the error is thrown prematurely. Using a setter on the object being assigned to is the only way around it I could find.

**Comment by nicolo-ribaudo** [comment]
Oh ok. Using a new helper looks good to me then (I'm working on almost the same thing but for private methods, and I didn't notice that `readOnlyError()._` doesn't work).

**Comment by overlookmotel** [comment]
Well it doesn't work in Node anyway.

Stripping out assignments which will never be made is a cleaner solution in terms of producing more minimal output:

Input:

```js
const c = 0;
let x, y;
[ x, c, y ] = [ 1, 2, 3 ];
```

Output:

```js
var c = 0;
var x, y;
var _c;
[ x, _c ] = [ 1, 2, 3 ], _readOnlyError("c");
```

This also works perfectly. It's just that it'd be a pain to implement handling when the object/array on the left is nested - `[ x, [ { y, c }, ...z ], q, v ]` etc.

**Comment by nicolo-ribaudo** [comment]
I prefer the new runtime helper: even if it generates more code at runtime, this is an edge case and I don't think it's worth the implementation complexity.

**Comment by overlookmotel** [comment]
OK great. I'll get onto that at some point. I'm afraid it may be a long time before I do though - really busy with work at the moment.

**Comment by overlookmotel** [comment]
By the way, I was intending to amend [plugin-transform-modules-commonjs](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs) to reuse the `_readOnlyError` helper. Currently it produces this function to throw an error inline every time.

However, I couldn't figure out how to access `state.addHelper()` deep within https://github.com/babel/babel/blob/main/packages/babel-helper-module-transforms/src/rewrite-live-references.ts Any ideas?

**Comment by nicolo-ribaudo** [comment]
It's a change that we should do, but the answer is not nice :stuck_out_tongue: 

You need to pass `state`/`file` as a parameter to `@babel/helper-module-transforms`, but also keep a fallback for when `state` is not passed (otherwise it would be a breaking change)

**Comment by overlookmotel** [comment]
Uck. I generally consider that the documentation is the public API contract. In this case, it is... minimal.

https://babeljs.io/docs/en/babel-helper-module-transforms

**Comment by nicolo-ribaudo** [comment]
It doesn't matter, every older version of `@babel/plugin-transform-modules-commonjs` would break if we add a new required parameter.
